discojs/src/models/gpt/hellaswag.ts

```diff
@@ -0,0 +1,177 @@
+import fs from 'fs';
+import { promises as fsPromises } from 'fs';
+import fetch from 'node-fetch';
+import * as tf from '@tensorflow/tfjs';
+import { GPT } from './index.js';
+import { tokenize } from '../../processing/text.js';
+import { PreTrainedTokenizer } from '@xenova/transformers';
+import * as readline from 'readline';
+import { fileURLToPath } from 'url';
+import path from 'path';
+import { List } from 'immutable';
+import { ONNXModel } from '../onnx.js';
+
+
+const HELLASWAG_URL = 'https://raw.githubusercontent.com/rowanz/hellaswag/master/data/hellaswag_val.jsonl';
+const __filename = fileURLToPath(import.meta.url);
+const __dirname = path.dirname(__filename);
+const LOCAL_FILE = path.resolve(__dirname, '../../../../datasets/hellaswag_val.jsonl');
+
+async function fileExists(path: string): Promise<boolean> {
+  try {
+    await fsPromises.access(path);
+    return true;
+  } catch {
+    return false;
+  }
+}
+
+export async function downloadHellaSwag(): Promise<void> {
+  if (await fileExists(LOCAL_FILE)) return;
+
+  const res = await fetch(HELLASWAG_URL);
+  const fileStream = fs.createWriteStream(LOCAL_FILE);
+
+  await new Promise<void>((resolve, reject) => {
+    res.body?.pipe(fileStream);
+    res.body?.on('error', reject);
+    fileStream.on('error', reject);
+    fileStream.on('finish', () => resolve());
+  });
+}
+
+interface HellaSwagExample {
+  ctx: string;
+  endings: string[];
+  label: number;
+}
+
+export async function* loadExamples(limit = 100): AsyncGenerator<HellaSwagExample> {
+  const fileStream = fs.createReadStream(LOCAL_FILE, 'utf-8');
+  const rl = readline.createInterface({ input: fileStream, crlfDelay: Infinity });
+
+  let count = 0;
+  for await (const line of rl) {
+    if (count++ >= limit) break;
+    try {
+      const data = JSON.parse(line.trim());
+      yield { ctx: data.ctx, endings: data.endings, label: data.label };
+    } catch (e) {
+      console.error(`Failed to parse line ${count}:`, line);
+      throw e;
+    }
+  }
+}
+
+async function computeLogLikelihood(gpt: GPT, inputIds: number[], ctxLength: number): Promise<number> {

```

Comment:
If this function (and computeONNXLogLikelihood) are adapted from Karparthy's code can you link his code as source?